index.rst

```diff
@@ -48,14 +48,177 @@
+Regardless, if the node is pre-empted during processing after files have been written, the orphaned files will be present and there will be no associated registry entries so attempting to clean-up does not gain us anything other than saving storage space if that run is never used again.
+The execution system is already capable of over-writing an existing file if it is not known to registry and datastore ensures that the file name is unique for a specific dataId and run combination.
+
+If pre-emption occurs during the final registry import it is assumed that there is a single transaction that will be completely rolled back.
+The end result would be that the entire processing would have to be redone.
+If the pre-emption occurs in the short time between completion of the registry import and the completion of the job, we should ensure that the ``pipetask`` command is configured to complete without action when the job is restarted.
+
+Externalized
+------------
+
+The most flexible approach is for the workflow generator to insert extra jobs into the workflow that handle registry merging and registry syncing.
+
+A new job would be inserted before the initial pipeline jobs in the graph that downloads the initial registry state from the cloud registry and stores it in a SQLite file.
+This SQLite file would then be provided as an input to the first job.
+The pipeline job would use that SQLite registry and a GCS-with-local-cache datastore as described in the previous section.
+On completion the output collected by the workflow execution system would be an updated SQLite registry.
+If the next pipeline job only takes files from a single input job this SQLite file would be passed directly to the next pipeline job.
+If, on the other hand, the next job gathers files from multiple jobs, the workflow generator will insert a new job before it that reads all the SQLite files and merges them into a single output for that job.
+It can optionally also sync registry information to the cloud registry and trim the SQLite file that it passes on.
+This sync and prune could be a distinct job that can be inserted after the merge job in the graph.
+At the end node (or nodes) of each workflow graph a final job will be inserted that updates the cloud registry with the final state.
+
+In this scenario pre-emption has no impact on registry state for pipeline jobs since they are running with an entirely local SQLite registry.
+Pre-emption still has to be understood in terms of the registry sync jobs and requires that the remote update of the cloud registry happens in a single transaction.
+
+This approach allows the submission system to decide whether the registry is updated multiple times within the graph or solely at the end since the registry merging jobs can be configured to either merge the inputs and pass them on complete, or merge, sync and trim.
+A trimmed registry can not be passed on to the next job if the remainder has not been synced with the cloud registry.
```

Comment:
This section seems to assume that there will be a single registry at the end guaranteed to have all of the data.  I think that is an option for this approach but not a requirement.  The end merge job could go through a bunch of registries being smart about only finding new data in each registries.   This last sentence in particularly tripped me up because I couldn't figure out why it was a requirement as the next job shouldn't be using the cloud registry.  I'm assuming that what is really meant it that the last/only copy of outputs cannot be trimmed unless it has been synced with the cloud registry.